internal/core/metadata/rest_deviceprofile_test.go

```diff
@@ -214,18 +230,42 @@ func TestAddDeviceProfile(t *testing.T) {
 				{"AddDeviceProfile", []interface{}{TestDeviceProfileValidated}, []interface{}{"", TestError}},
 				{"GetDeviceProfileByName", []interface{}{TestDeviceProfileName}, []interface{}{contract.DeviceProfile{}, db.ErrNotFound}},
 			}),
-			MockValueDescriptorClient{},
+			MockValueDescriptorClient{&mock.Mock{}, nil},
+			true,
+			true,
 			http.StatusInternalServerError,
 		},
+		{
+			"Add device profile without enable value descriptor management",
+			createRequestWithBody(http.MethodPost, TestDeviceProfile),
+			createDBClientWithOutlines([]mockOutline{
+				{"AddDeviceProfile", []interface{}{TestDeviceProfileValidated}, []interface{}{TestDeviceProfileID, nil}},
+				{"GetDeviceProfileByName", []interface{}{TestDeviceProfileName}, []interface{}{contract.DeviceProfile{}, db.ErrNotFound}},
+			}),
+			MockValueDescriptorClient{&mock.Mock{}, nil},
+			false,
+			false,
+			http.StatusOK,
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			rr := httptest.NewRecorder()
 			var loggerMock = logger.NewMockClient()
 			configuration := metadataConfig.ConfigurationStruct{
-				Writable: metadataConfig.WritableInfo{EnableValueDescriptorManagement: true},
+				Writable: metadataConfig.WritableInfo{EnableValueDescriptorManagement: tt.enableValueDescriptorManagement},
 				Service:  bootstrapConfig.ServiceInfo{MaxResultCount: 1},
 			}
+			if tt.valueDescriptorAddMethodCalled {
+				for _, dr := range TestDeviceProfile.DeviceResources {
+					vd := contract.From(dr)
+					tt.vdcMock.On("Add", tt.request.Context(), &vd).Return("", nil).Once()
+					// Value descriptor client occur error, so only call the add method once.
+					if tt.vdcMock.errorToThrow != nil {
+						break
+					}
+				}
+			}
```

Comment:
Can we setup the `ValueDescriptorMock` in a separate function which we can call during the test arguments, like we are doing with the `dbClient`. My worry is that this block complicates the test in a way that affects tests that are not using this functionality, so by keeping it contained in the test argument we keep things simple and easier to maintain. I would suggest creating a function like `createDBClientWithOutlines`. This will allow us to leverage the fact that the mock will panic if a method is invoked that was not setup to be mocked. Therefore making the tests easy to verify, since for cases we do not expect it to be called would result in a panic and the test will naturally fail, and for cases it should be called we can rely on the mock has to be invoked with the exact arguments it was setup to accept. Finally in the test assertion section we can do a simple test that if `test.ExpectValueDescritorCreated`(We would need to add this test argument) is true then we can ensure the mock was called one time using one of the `assert.Called...` functions.

This applies to the `TestAddProfileByYaml` test as well. I hope this makes sense.

Reply:
@AnthonyMBonafide No problem, testing modified, please help review.